src/digraph-cyclicity.jl

```diff
@@ -0,0 +1,439 @@
+* `dg`: the digraph from which cycles are computed
+* `visitor`: Informations needed for the cycle computation, contains:
+    * `stack`: the stack of parent vertices
+    * `blocked`: tells whether a vertex has already been explored or not
+    * `blockedmap`: mapping of the blocking / unblocking consequences
+* `allcycles`: output containing the cycles already detected
+* `vmap`: vector map containing the link from the old to the new nodes of the directed graph
+* `startnode = v`: optional argument giving the starting node. In the first iteration,
+ the same as v, otherwise it should be passed.
+
+# Returns
+* `done`: tells whether a circuit has been found in the current exploration.
+"""
+function circuit(v::Int, dg::DiGraph, vis::JohnsonVisitor, allcycles::Vector{Vector{Int}}, vmap::Vector{Int}, startnode = v)
+    done = false
+    push!(vis.stack, v)
+    vis.blocked[v] = true
+    for w in fadj(dg,v)
+        if w == startnode
+            push!(allcycles, vmap[vis.stack])           
+            done = true
+        elseif !vis.blocked[w]
+            circuit(w, dg, vis, allcycles, vmap, startnode) && (done = true)
+        end
+    end
+    if done
+        unblock!(v, vis.blocked, vis.blockedmap)
+    else
+        for w in fadj(dg, v)
+			if !in(vis.blockedmap[w], v)
+				push!(vis.blockedmap[w], v)
+			end
+        end
+    end
+    pop!(vis.stack)
+    return done
+end
+
+
+"""
+```simplecycles(dg::DiGraph)```
+
+Compute all cycles of the given directed graph, using 
+[Johnson's algorithm](http://epubs.siam.org/doi/abs/10.1137/0204007). 
+
+/!\ The number of cycles grow more than exponentially with the number of vertices, 
+you might want to use the algorithm with a ceiling -- `getcycles` -- on large directed graphs 
+(slightly slower). If you want to have an idea of the possible number of cycles, 
+look at function ```maxcycles(dg::DiGraph, byscc::Bool = true)```.
+
+# Arguments
+* `dg`: the directed graph
+
+# Returns
+* `cycles`: all the cycles of the directed graph
+"""
+function simplecycles(dg::DiGraph)
+    sccs = strongly_connected_components(dg)
+    cycles = Vector{Vector{Int}}()
+    for scc in sccs
+        #while length(scc) > 1
+        for i in 1:(length(scc)-1)
+            wdg, vmap = induced_subgraph(dg, scc[i:end])
+            #startnode = 1
+            #shift!(scc)
+            visitor = JohnsonVisitor(wdg)
+            circuit(1, wdg, visitor, cycles, vmap)
+        end
+    end
+    return cycles
+end
+
+
+##########################################################
+#### Iterative version, using Tasks, of the previous algorithms.
+"""
+```circuit(v::Int, dg::DiGraph, vis::JohnsonVisitor, vmap::Vector{Int}, startnode = v)```
+
+One step of the recursive version of simple cycle detection, using a DFS algorithm.
+
+The CIRCUIT function from [Johnson's algorithm](http://epubs.siam.org/doi/abs/10.1137/0204007),
+ recursive and iterative version. Produce a cycle when needed, can be used only inside a 
+ Task.
```

Comment:
Thanks for the remarks! My understanding of iterators and Tasks in Julia is basic, so I might completely be wrong. If I understand you correctly however, here is the answer:
There is an iterative version of the function a few lines before, used in the simplecycles function. It does not spit out an iterator however (I don't know how to do that properly).
When I coded Johnson's algorithm, the use of tasks has one big purposes, reduce memory consumption, because the original version (in the simplecycles function) can be extremely memory intensive with large graphs. The reduction in memory consumption is made with two different tricks:
- stopping the iteration at a given ceiling, so that if the graph is too large, only a subsample of all cycles are spit out, reducing memory usage.
- I think I create some kind of iterator with this method.
I used this because of my limited knowledge in Julia. If there are better ways, I am completely open to it, I just do not know how to do it yet.